Mockito plugin annotation

mat013 opened the pull request.

mat013 commented:
I found another issue

The class GlobalConfiguration contains a ThreadLocal which contains the latest global configuration, and when running the test the TestBase invokes ConfigurationAccess.getConfig which casts the GlobalConfiguration to a MockitoConfiguration, which is not the case when using the @MockitoPlugin. 

I guess the GlobalConfiguration is meant to serve as a singleton, however I guess that is not possible when you have a annotation engine for each test instance.

Are there any special reason why global configuration has to use a ThreadLocal ?

bric3 commented on src/main/java/org/mockito/internal/configuration/GlobalConfiguration.java:
```diff
@@ -25,6 +26,10 @@
+    //back door for testing
+    public static void removeIt() {
+        GLOBAL_CONFIGURATION.remove();
+    }
```
Don't add this method here, for unit tests there's this class `org.mockito.configuration.MockitoConfiguration` in the test sources

bric3 commented on src/test/java/org/mockitousage/MockitoAnnotationsTest.java:
```diff
@@ -0,0 +1,89 @@
+    
+    @RunWith(MockitoJUnitRunner.class)
+    @MockitoPlugin(annotationEngine=MyAnnotationEngine.class)
+    public static class MockitoPluginAnnotationTestClass {
```
I think such annotations shouldn't be on the test class, but rather as a class field (or maybe both), so a plugin could be  be instantiated with custom code (like parameters, other dependencies, etc.)

mat013 replied:
I think this can be achieved by adding a custom the annotation engine on class level.
I think if it is needed to have an annotation for class fields then we can build a annotation engine which could support the new annotation for the class field. 

What do you think?

bric3 commented on src/test/java/org/mockitousage/MockitoAnnotationsTest.java:
```diff
@@ -0,0 +1,89 @@
+        NoMockitoPluginAnnotationTestClass noMockitoPluginAnnotationTestClass = new NoMockitoPluginAnnotationTestClass(); 
+
+        MockitoAnnotations.initMocks(noMockitoPluginAnnotationTestClass);
+        assertThat(isMock(noMockitoPluginAnnotationTestClass.someMock), is(true));
```
You may want to prefer AssertJ assertions, it's in our dependencies (and it replaces FEST Assert).

bric3 commented:
> Are there any special reason why global configuration has to use a ThreadLocal ?

I don't remember. @szczepiq do you?

Nice feature overall :)

bric3 commented on src/main/java/org/mockito/MockitoPlugin.java:
```diff
@@ -0,0 +1,73 @@
+@Retention(RetentionPolicy.RUNTIME)
+@Target(ElementType.TYPE)
+@Inherited
+public @interface MockitoPlugin {
```
Open thinking mode, I wonder it should be named `MockitoAnnotationEngine`, instead of adding multiple fields here for each new mockito plugin

bric3 commented:
By the way your commits should be squashed ;)

mat013 commented:
> By the way your commits should be squashed ;)

I tried let's see if I did it correctly :p

mat013 commented:
I moved the removeIt from the globalconfiguration to ConfigurationAccess instead, I think there is an issue with the global configuration that it can only be instantiated once. thereby it will not be possible to have different annotation engine for different test cases.

I think it would make sense to do that so I change the new constructor in GlobalConfiguraton which I implemented to _always_ instantiate a new configuration with the provided annotation engine.

bric3 commented:
>> By the way your commits should be squashed ;)
> I tried let's see if I did it correctly :p

Nope, you have imported commits from master , and you somehow did some merge commits from another branch. On _pull request_ branches it is better to keep a clean _flat_ history.

I'm not sure why you needed those merges, but you definitely want to avoid those. _Squashing_ means to put commits together in order to have a clean and meaningful history. This is done via a `git rebase --interactive <the commit references>`. Check our [wiki](https://github.com/mockito/mockito/wiki/Using-git-to-prepare-your-PR-to-have-a-clean-history) on that matter, it contains the bare minimum to understand how to have a clean history. Nevertheless there's better and more complete tutorial out there.

Don't worry I know git is a bit hard when beginning, but once you know your way and why it matters, you won't regret it :)

mat013 commented:
So should I make a new branch and transfer all the changes into the new
branch rather than continue working on this messy branch?

Br

Meang

On Sun, Jul 12, 2015, 19:50 Brice Dutheil <notifications@github.com> wrote:

> By the way your commits should be squashed ;)
>
> I tried let's see if I did it correctly :p
>
>  Nope, you have imported commits from master , and you somehow did some
> merge commits from another branch. On *pull request* branches it is
> better to keep a clean *flat* history.
>
> I'm not sure why you needed those merges, but you definitely want to avoid
> those. *Squashing* means to put commits together in order to have a clean
> and meaningful history. This is done via a git rebase --interactive <the
> commit references>. Check our wiki
> <https://github.com/mockito/mockito/wiki/Using-git-to-prepare-your-PR-to-have-a-clean-history>
> on that matter, it contains the bare minimum to understand how to have a
> clean history. Nevertheless there's better and more complete tutorial out
> there.
>
> Don't worry I know git is a bit hard when beginning, but once you know
> your way and why it matters, you won't regret it :)
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mockito/mockito/pull/255#issuecomment-120750213>.
>

bric3 commented:
Yes you can.
Can you close this one when you are ready :)

mat013 commented:
Killed it as it the branch was messed up. Have created a new PR and branch for that feature
